nemo_aligner/utils/utils.py

```diff
@@ -123,7 +123,34 @@ def load_and_override_model_config(restore_path, model_cfg_to_overwrite, remove_
         checkpoint_cfg.pop("target", None)
         checkpoint_cfg.pop("nemo_version", None)
 
-    return OmegaConf.merge(checkpoint_cfg, model_cfg_to_overwrite)
+    resolved_checkpoint_cfg = resolve_overlapping_fields(checkpoint_cfg, model_cfg_to_overwrite)
+
+    return OmegaConf.merge(resolved_checkpoint_cfg, model_cfg_to_overwrite)
+
+
+def resolve_overlapping_fields(base_config, overwrite_config):
+    def recursive_merge(base, overwrite, path=[]):
+        for key, value in overwrite.items():
+            current_path = path + [key]
+
+            if key not in base:
+                continue
+
+            if isinstance(value, DictConfig) and isinstance(base[key], DictConfig):
+                recursive_merge(base[key], value, path=current_path)
+            else:
+                current_config = overwrite_base_config
+                for sub_key in current_path:
+                    if isinstance(current_config, DictConfig):
+                        current_config = current_config[sub_key] if sub_key in current_config else None
```

Comment:
I'm a bit confused by what this `current_config` is doing. Since we're going down the same path we went through aren't we guaranteed that `overwrite_base_config` has such a path? And isn't `current_config` essentially the same as `value`?

Reply:
Resolving this comment as I had misunderstood this code.